Updated API design

Pragadeesh-19 opened the pull request:
Submitting updated API design yaml file for reivew.

Pragadeesh-19 committed:
feat: Added Swagger API design yaml file

Pragadeesh-19 committed:
updated code/api-definition/dynamodb

shubhamkumar-cyg commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,1011 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  /clients/{clientId}/therapists/{therapistId}/messages:
```
merge client and therapist message endpoint into one endpoint /messages and pass data in request body

shubhamkumar-cyg commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,1011 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  /clients/{clientId}/therapists/{therapistId}:
```
move everything related to client therapist interactions like access request, listing client/therapist etc inside /relationships endpoint

shubhamkumar-cyg commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,1011 @@
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/StandardResponse'
+  /clients/search:
```
merge client and therapist search endpoints into one /search

shubhamkumar-cyg commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,1011 @@
+            application/json:
+              schema:
+                 $ref: '#/components/schemas/StandardResponse'
+  /clients/{clientId}/emotions:
```
add all the things related to journal inside /emotion-journal endpoint

shubhamkumar-cyg commented on swagger/swagger.yml:
Looks very good.

I have given few comments that will help cleaning up the API by grouping similar things together into fewer, more organized endpoints. This makes it simpler to use and manage.

shubhamkumar-cyg reviewed (changes requested):
Looks very good.

I have given few comments that will help cleaning up the API by grouping similar things together into fewer, more organized endpoints. This makes it simpler to use and manage.

Pragadeesh-19 committed:
Refractor API design with relationship grouping and parameter updates

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  
```
`/relationship` endpoints for client-therapist interaction.

Pragadeesh-19 commented:
Updated the API design based on the feedback and add the following:
1. added emotion and journal related under `/emotion-journal/`.
2. merged search endpoints of client and therapist under `/search`. 
3. moved client-therapist relationship under `/relationship`. 
4. merged client-therapist messaging endpoints under `/messages`.

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+    get:
+      summary: Get the list of therapists or clients based on user type
+      security:
+        - bearerAuth: []
```
`/relationship` endpoint for client-therapist endpoint.

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+  
+  /relationships:
+    get:
+      summary: Get the list of therapists or clients based on user type
```
`/relationship` for client-therapist interaction

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+                $ref: '#/components/schemas/StandardResponse'    
+  /messages:
+    post:
+      summary: Send a message to a therapist or client
```
merged client-therapist messaging endpoints under `/messages`

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+                $ref: '#/components/schemas/StandardResponse'
+  /emotion-journal:
+    post:
+      summary: Add a new emotion record
```
added journal related endpoints under `/emotion-journal`

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  
+  /relationships:
```
`/relationship` for client-therapist interaction

cygnius replied:
Update and read APIs are missing for this object.

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'    
+  /messages:
```
merged client-therapist messaging endpoints under `/messages`

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  /emotion-journal:
```
added journal related endpoints under `/emotion-journal`

cygnius replied:
clientId should be in path of this endpoint.

Pragadeesh-19 commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/StandardResponse'
+  /search:
```
merged search endpoints of client and therapist under `/search`

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            - CANCELLED
+
+paths:
+  /auth/register:
```
When you have a Create API for object X (for example client) which has attribute A,B,C, create two schema objects - XCreationInput and X.
X schema will have attributes Xid, A,B,C. XCreationInput will have attributes A,B,C.
Use XCreationInput schema object reference to specify the body of Create API request.
Use X schema object reference to specify return type of Create/Update/Update/List API. 
Reasons: Id of the object is never specified by the client. It is decided at the backend server side to avoid any collisions/hack attempts. Hence it is not present in the request of the API call and present in the response. Client MUST get the id of the object when calling Create/Read/Update/List API to that it can power other necessary parts of the app. 
Implement this everywhere you have a “Create” API.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            - error
+        message:
+          type: string
+        data:
+          type: object
```
This is too generic response object. 
It defeats the purpose of specifying all details in the Swagger YAML. 
API YAML specifies input and output object at API level so that the understanding between front end and backend developers is same. It will save us the time of writing model and validation code for these response and request object on client and server side.
There should be schema objects for each type of response object - Client, Therapist, Appointment, etc.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
```
Response should have the generated user id of the user in the system.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
```
Clarify the schema of the response.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              - THERAPIST
+      responses:
+        '200':
+          description: Successfully retrieved list of mapped therapists
```
Fix the description.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+        '201':
+          description: Therapist mapped to client successfully
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
```
Clarify the response object.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
```
Clarify the response object. This feedback is for each and every API.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  /relationships/journalAccessRequests:
```
This doesnt need to be a separate endpoint. You can merge it with /relationships/journalAccess endpoint.

You can introduce a attribute called "STATE" in the journalAccess object. This attribute is in "requested" state first.
Once it is approved (Update API), it is changed to "APPROVED".

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  /relationships/journalAccess:
```
Update, Read and List APIs are missing for this endpoint.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+      security:
+        - bearerAuth: []
+      parameters:
+        - in: query
+          name: clientId
+          required: true
+          schema:
+            type: string
+        - in: query
+          name: therapistId
+          required: true
+          schema:
```
Why not pass it in the body itself?

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+    get:
+      summary: Get the history of messages 
+      security:
+        - bearerAuth: []
+      parameters:
+        - in: query
+          name: senderId
+          required: true
+          schema:
+            type: string
+        - in: query
+          name: receiverId
+          required: true
+          schema:
+            type: string
```
What about messages which were "sent" by the "receiver" and "received" by "sender". Essentially replies to the messages?

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+                properties:
+                  therapistId:
+                    type: string
+                  clientId:
+                    type: string
```
Therapist creates "Slot/session" with date, start & end time showing their availability.
Then client will list available sessions for a therapist and click and request to book one.
This request will be approved/rejected by the Therapist.
Structure your APIs around this flow.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'
+  /appointments:
```
therapistId should be in the path because therapist "owns" the session.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              application/json:
+                schema:
+                   $ref: '#/components/schemas/StandardResponse'
+    get:
```
We need the API also to return -  Get the list of appointments requests routed to a therapist.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/StandardResponse'
+  /therapists/{therapistId}/appointments:
```
Doesn't need to be a separate path. It can be a part of the same path before this.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+          required: true
+          schema:
+            type: string
+        - in: query
+          name: userType
+          required: true
+          schema:
+            type: string
+            enum:
+              - CLIENT
+              - THERAPIST
```
Why did you pass this? The server would already know the type of the user who is calling the API.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/StandardResponse'       
+  /therapists/{therapistId}/sessions:
```
This can be merged with appointment's path. Where the user is putting private and shared notes against a booked session.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+              application/json:
+                schema:
+                   $ref: '#/components/schemas/StandardResponse'
+    get:
```
can be merged with appointment's path.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                 $ref: '#/components/schemas/StandardResponse'
+  /therapists/{therapistId}/sessions/{sessionId}:
```
can be merged with appointment's path.

cygnius commented on swagger/swagger.yml:
```diff
@@ -0,0 +1,831 @@
+            application/json:
+              schema:
+                 $ref: '#/components/schemas/StandardResponse'
+    get:
```
can be merged with appointment's path.

cygnius commented:
Overall Review:
1) You have not clearly identified the objects and their attributes, request and response objects.
2) The APIs are mixed across different paths and need to be consolidated.

Think about each real world objects and its attributes first. How they are related with each other. How and who creates them, reads them, updates them in what order. This clarity will help you structure the APIs better.
Go through individual comments to understand.
